File: terraform/functions/kinesis_twitter_table.js
const axios     = require('axios')
const AWS       = require('aws-sdk')
const converter = AWS.DynamoDB.Converter.unmarshall

const host    = process.env.ENDPOINT
const index   = process.env.INDEX
const url     = `https://${host}/${index}/_doc`
const headers = { 'Content-Type': 'application/json' }

exports.handler = async event => {
    await Promise.all(event.Records.map(record => {
        const partitionKey = record.dynamodb.Keys.partitionKey.S
        const rangeKey = record.dynamodb.Keys.rangeKey.S
        const itemUrl = `${url}/${partitionKey}_${rangeKey}`

        if (record.eventName === 'REMOVE') {
            return axios.delete(itemUrl)
        } else {
            let document = converter(record.dynamodb.NewImage)
            document = removeTechnicalFields(document)
            return axios.put(
                itemUrl,
                JSON.stringify(document),
                { headers }
            )
        }
    }))
}

const removeTechnicalFields = (body) => {
  delete body.SequenceNumber
  delete body.SizeBytes
  delete body.StreamViewType
  delete body['aws:rep:deleting']
  delete body['aws:rep:updatetime']
  delete body['aws:rep:updateregion']
  return body
}

File: terraform/functions/kinesis_twitter_table.py
import boto3
import requests
from requests_aws4auth import AWS4Auth
import os

#From aws documentation
region = 'us-west-2' # e.g. us-east-1
service = 'es'
credentials = boto3.Session().get_credentials()
awsauth = AWS4Auth(credentials.access_key, credentials.secret_key, region, service, session_token=credentials.token)

host = f"https://{os.environ['ENDPOINT']}"  # the OpenSearch Service domain, e.g. https://search-mydomain.us-west-1.es.amazonaws.com
index = os.environ['INDEX']
datatype = '_doc'
url = host + '/' + index + '/' + datatype + '/'

headers = { "Content-Type": "application/json" }

def handler(event, context):
    count = 0
    for record in event['Records']:
        # Get the primary key for use as the OpenSearch ID
        id = record['dynamodb']['Keys']['id']['S']

        if record['eventName'] == 'REMOVE':
            r = requests.delete(url + id, auth=awsauth)
        else:
            document = record['dynamodb']['NewImage']
            r = requests.put(url + id, auth=awsauth, json=document, headers=headers)
        count += 1
    return str(count) + ' records processed.'